### ISSUE TITLE
Models are not saved after training.

### ISSUE THREAD

**Issue by sainimohit23** [as-ideas/headliner#16]
I am training a transformer model on kaggle. After the training, models are not getting saved. Earlier it was working fine before the addition of BERT transformer. 

Here is the notebook: https://www.kaggle.com/mohitsaini235/chatbot?scriptVersionId=22931537

I have commented out some code so that you can run it quickly and see the results.

**Comment by cschaefer26** [comment]
Hi, I just tested model save/load and saw now issues. Could you try to change the model save path from /summarizer to /tmp/summarizer? Could be a permission issue.

**Comment by sainimohit23** [comment]
@cschaefer26 I don't know if it has something to do with the kaggle. But, just adding a forward slash in front of save_path worked. Ah! life with a potato pc. Anyway, thank you for helping me out. I love this library.

**Comment by sainimohit23** [comment]
@cschaefer26 I checked it again with different paths. Models are not saving when I am using the custom preprocessing. Without preprocessing the trainer is saving the checkpoints. This is the code:

```
preprocessor = Preprocessor(filter_pattern='(["#$%&()*+/:;<=>@[\\]^`{|}~\t\n])',
                            punctuation_pattern='([!.?,])',
                            lower_case=True,
                            hash_numbers=False)
train_prep = [preprocessor(t) for t in train]
inputs_prep = [t[0] for t in train_prep]
targets_prep = [t[1] for t in train_prep]


from tensorflow_datasets.core.features.text import SubwordTextEncoder
tokenizer_input = SubwordTextEncoder.build_from_corpus(
inputs_prep, target_vocab_size=2**13, reserved_tokens=[preprocessor.start_token, preprocessor.end_token])
tokenizer_target = SubwordTextEncoder.build_from_corpus(
    targets_prep, target_vocab_size=2**13,  reserved_tokens=[preprocessor.start_token, preprocessor.end_token])


vectorizer = Vectorizer(tokenizer_input, tokenizer_target)
summarizer = SummarizerTransformer(max_prediction_len=20, embedding_size=EMB_SIZE)
summarizer.init_model(preprocessor, vectorizer)

trainer = Trainer(batch_size=BATCH_SIZE, 
                  steps_per_epoch=STEPS_PER_EPOCH, 
                  steps_to_log=STEPS_TO_LOG, 
                  max_output_len=MAX_OUTPUT_LENGTH, 
                  model_save_path=save_path)
trainer.train(summarizer, train, num_epochs=1, callbacks=[CustomCallback()])

```

**Comment by cschaefer26** [comment]
Thx for the minimal example, I'm going to check it right now.

**Comment by cschaefer26** [comment]
Hi, one thing I notice is that you don't provide validation data:

trainer.train(summarizer, train, val_data=val)

Without val data the trainer is not saving the model (since by default the checkpoint is looking at the val score). If you don't have validation data, you can simply use a subset of the training data:

trainer.train(summarizer, train, val_data=train). 

Could you check if thats resolving it?

**Comment by sainimohit23** [comment]
@cschaefer26 It worked. It is not clear from the documentation that the model will be saved only If the validation loss is improved.

**Comment by cschaefer26** [comment]
Updated to 0.0.22 - models are now saved even when no validation data is provided.
